API/RecommendationService/offline_data_service.py

```diff
@@ -1,41 +1,45 @@
+import asyncio
 import os
 
 from azure.cosmos import CosmosClient, PartitionKey
 from .util import get_latest_cmd, generated_query_kql, RecommendationSource, RecommendType, generated_cosmos_type, CosmosType
 
 
-def get_recommend_from_offline_data(command_list, recommend_type, error_info, top_num=50):
+async def get_recommend_from_offline_data(command_list, recommend_type, error_info, top_num=50):
+    loop = asyncio.get_event_loop()
     cosmos_type = generated_cosmos_type(recommend_type, error_info)
     commands = get_latest_cmd(command_list, 2)
 
     client = CosmosClient(os.environ["CosmosDB_Endpoint"], os.environ["CosmosDB_Key"])
-    database = client.create_database_if_not_exists(id=os.environ["CosmosDB_DataBase"])
+    database = client.get_database_client(os.environ["CosmosDB_DataBase"])
     totalcount_threshold = int(os.environ["Solution_TotalCount_Threshold"]) if cosmos_type == CosmosType.Solution else int(os.environ["Command_TotalCount_Threshold"])
     ratio_threshold = int(os.environ["Solution_Ratio_Threshold"]) if cosmos_type == CosmosType.Solution else int(os.environ["Command_Ratio_Threshold"])
 
     if cosmos_type == CosmosType.Solution:
         return get_recommend_from_cosmos(database, commands[-1:], recommend_type, error_info, totalcount_threshold, ratio_threshold, top_num)
     else:
         # The recommended content matching the last two commands is preferred. If there is no data, it will fall back to the situation of matching the last command
-        result = get_recommend_from_cosmos(database, commands[-2:], recommend_type, error_info, totalcount_threshold, ratio_threshold, top_num)
-        if len(result) > 0:
-            return result
+        result_2_future = loop.run_in_executor(None, get_recommend_from_cosmos, database, commands[-2:], recommend_type, error_info, totalcount_threshold, ratio_threshold, top_num)
+        result_future = loop.run_in_executor(None, get_recommend_from_cosmos, database, commands[-1:], recommend_type, error_info, totalcount_threshold, ratio_threshold, top_num)
 
-        return get_recommend_from_cosmos(database, commands[-1:], recommend_type, error_info, totalcount_threshold, ratio_threshold, top_num)
+        result_2 = await result_2_future
+        if len(result_2) > 0:
+            return result_2
+        else:
+            return await result_future
```

Comment:
Since both `result_2_future` and `result_future` will be queried at the same time, perhaps we can consider using `result_future` to supplement if the quantity of `result_2_future` is insufficient

Reply:
> Since both `result_2_future` and `result_future` will be queried at the same time, perhaps we can consider using `result_future` to supplement if the quantity of `result_2_future` is insufficient

I agree.

Reply:
Done